service/controller/azureconfig/handler/volumebindingmigration/resource.go

```diff
@@ -0,0 +1,179 @@
+
+		var presentObj storagev1.StorageClass
+		err := tenantClusterK8sClient.Get(ctx, ctrl.ObjectKey{Name: desiredObj.Name, Namespace: desiredObj.Namespace}, &presentObj)
+		if apierrors.IsNotFound(err) {
+			// All good. We'll create it.
+			r.logger.Debugf(ctx, "did not find present storage class object %q", desiredObj)
+		} else if err != nil {
+			return microerror.Mask(err)
+		} else {
+			r.logger.Debugf(ctx, "finding if present storage class object %q has desired volumeBindingMode %q", desiredObj.Name, desiredObj.VolumeBindingMode)
+		}
+
+		// If present object matches the desired one, continue to next one.
+		if presentObj.VolumeBindingMode != nil && *presentObj.VolumeBindingMode == *desiredObj.VolumeBindingMode {
+			r.logger.Debugf(ctx, "present storage class object %q has desired volumeBindingMode: %q", presentObj.Name, desiredObj.VolumeBindingMode)
+			continue
+		}
+
+		// Volume bind mode is immutable field so we must delete the present
+		// object if it exists.
+		if !presentObj.CreationTimestamp.IsZero() && presentObj.ResourceVersion != "" {
+			r.logger.Debugf(ctx, "present storage class object %q does not have desired volumeBindingMode but %q instead", presentObj.Name, presentObj.VolumeBindingMode)
+			r.logger.Debugf(ctx, "deleting present storage class object %q", presentObj.Name)
+			err = tenantClusterK8sClient.Delete(ctx, &presentObj)
+			if err != nil {
+				return microerror.Mask(err)
+			}
+			r.logger.Debugf(ctx, "deleted present storage class object %q", presentObj.Name)
+		}
+
+		r.logger.Debugf(ctx, "creating desired storage class object %q", desiredObj.Name)
+
+		// Finally create the desired object.
+		err = tenantClusterK8sClient.Create(ctx, &desiredObj)
+		if err != nil {
+			return microerror.Mask(err)
+		}
+
+		r.logger.Debugf(ctx, "created desired storage class object %q", desiredObj.Name)
+	}
+
+	r.logger.Debugf(ctx, "ensured storageclasses use desired volumeBindingMode")
+
+	return nil
+}
+
+// EnsureDeleted is no-op.
+func (r *Resource) EnsureDeleted(ctx context.Context, obj interface{}) error {
+	return nil
+}
+
+// Name returns the resource name.
+func (r *Resource) Name() string {
+	return Name
+}
+
+func (r *Resource) getTenantClusterClient(ctx context.Context, azureConfig *providerv1alpha1.AzureConfig) (ctrl.Client, error) {
+	var k8sClient k8sclient.Interface
+	{
+		restConfig, err := r.tenantRestConfigProvider.NewRestConfig(ctx, key.ClusterID(azureConfig), key.ClusterAPIEndpoint(*azureConfig))
+		if err != nil {
+			return nil, microerror.Mask(err)
+		}
+
+		k8sClient, err = k8sclient.NewClients(k8sclient.ClientsConfig{
+			Logger:     r.logger,
+			RestConfig: rest.CopyConfig(restConfig),
+		})
+		if err != nil {
+			return nil, microerror.Mask(err)
+		}
+	}
+
+	return k8sClient.CtrlClient(), nil
+}
+
+func defaultStorageClasses() []storagev1.StorageClass {
+	var storageClasses []storagev1.StorageClass
+
+	objs := bytes.Split([]byte(ignition.DefaultStorageClass), []byte("---"))
```

Comment:
I feel like this is super stupid way to do this, but there's no streaming parsing implemented in k8s YAML parser and none others works correctly with the object meta annotations :man_facepalming:

Reply:
good enought for me, apart from the `panic(err)`

Reply:
You think it should be a handled error despite of built-in templating?

Reply:
I don't know, will leave that to you.
But the panic feels wrong to me in general.

Reply:
I added extra error handling there.

Reply:
thx